### PR TITLE
Add "Can form a Quorum" indicator in Health API

### PR DESCRIPTION
**Pull request by arteam**
Based on #83119 and #83205

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/coordination/indicators/LackOfQuorumIndicator.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        return CLUSTER_COORDINATION;
+    }
+
+    @Override
+    public HealthIndicatorResult calculate() {
+        var clusterState = clusterService.state();
+        boolean hasQuorum = clusterState.getLastCommittedConfiguration()
+            .hasQuorum(clusterState.getNodes().getNodes().values().stream().map(DiscoveryNode::getId).collect(Collectors.toSet()));
+        return hasQuorum
+            ? HealthIndicatorResult.of(LACK_OF_QUORUM, CLUSTER_COORDINATION, HealthStatus.GREEN, "Instance can form a quorum.")
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/coordination/indicators/LackOfQuorumIndicator.java:46]:
```suggestion
            ? HealthIndicatorResult.of(LACK_OF_QUORUM, CLUSTER_COORDINATION, HealthStatus.GREEN, "Instances can form a quorum.")
```

File: server/src/main/java/org/elasticsearch/cluster/coordination/indicators/LackOfQuorumIndicator.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            : HealthIndicatorResult.of(
+                LACK_OF_QUORUM,
+                CLUSTER_COORDINATION,
+                HealthStatus.RED,
+                "Not enough nodes with voting rights to form a quorum."
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/coordination/indicators/LackOfQuorumIndicator.java:51]:
```suggestion
                "Not enough master eligible instances to form a quorum."
```

Should we also list detected instances in the details?

File: server/src/main/java/org/elasticsearch/cluster/coordination/indicators/LackOfQuorumIndicator.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.health.HealthIndicatorResult;
+import org.elasticsearch.health.HealthIndicatorService;
+import org.elasticsearch.health.HealthStatus;
+
+import java.util.stream.Collectors;
+
+public class LackOfQuorumIndicator implements HealthIndicatorService {
+
+    public static final String LACK_OF_QUORUM = "lack_of_quorum";
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/coordination/indicators/LackOfQuorumIndicator.java:21]: `lack_of_quorum: green` <- lack_of_quorum is bad, but green is good. I am confused.
May be `can_form_quorum`?

File: server/src/main/java/org/elasticsearch/cluster/coordination/indicators/LackOfQuorumIndicator.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    public static final String CLUSTER_COORDINATION = "cluster_coordination";
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/coordination/indicators/LackOfQuorumIndicator.java:22]: You could use common `ServerHealthComponents.CLUSTER_COORDINATION` constant

File: server/src/main/java/org/elasticsearch/cluster/coordination/indicators/LackOfQuorumIndicator.java (hunk 2 above)
Comment (arteam) [reply server/src/main/java/org/elasticsearch/cluster/coordination/indicators/LackOfQuorumIndicator.java:51]: https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-node.html#voting-only-node

Voting-only nodes also can participate in a quorum.

File: server/src/main/java/org/elasticsearch/cluster/coordination/indicators/LackOfQuorumIndicator.java (hunk 3 above)
Comment (arteam) [reply server/src/main/java/org/elasticsearch/cluster/coordination/indicators/LackOfQuorumIndicator.java:21]: Very good point! I've renamed the class and the indicator.

File: server/src/main/java/org/elasticsearch/cluster/coordination/indicators/LackOfQuorumIndicator.java (hunk 2 above)
Comment (arteam) [reply server/src/main/java/org/elasticsearch/cluster/coordination/indicators/LackOfQuorumIndicator.java]: But they are also master-eligible, so let's me apply your suggestion here!

**Review by idegtiarenko — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/cluster/coordination/indicators/CanFormQuorumIndicator.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    @Override
+    public String component() {
+        return CLUSTER_COORDINATION;
+    }
+
+    @Override
+    public HealthIndicatorResult calculate() {
+        var clusterState = clusterService.state();
+        boolean hasQuorum = clusterState.getLastCommittedConfiguration()
+            .hasQuorum(clusterState.getNodes().getNodes().values().stream().map(DiscoveryNode::getId).collect(Collectors.toSet()));
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/coordination/indicators/CanFormQuorumIndicator.java:45]: I am new to this api. Is it returning all nodes or additional filtering for master/voting role is required?

Comment (arteam) [reply server/src/main/java/org/elasticsearch/cluster/coordination/indicators/CanFormQuorumIndicator.java]: Good catch, it got mangled when extracting to a separate PR, it should be `clusterState.getNodes().getMasterNodes()`

File: server/src/main/java/org/elasticsearch/cluster/coordination/indicators/CanFormQuorumIndicator.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import org.elasticsearch.health.HealthIndicatorService;
+import org.elasticsearch.health.HealthStatus;
+
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.health.ServerHealthComponents.CLUSTER_COORDINATION;
+
+public class CanFormQuorumIndicator implements HealthIndicatorService {
+
+    public static final String LACK_OF_QUORUM = "can_form_quorum";
```
Comment (tbrooks8) [inline server/src/main/java/org/elasticsearch/cluster/coordination/indicators/CanFormQuorumIndicator.java:23]: Maybe either name this constant NAME or CAN_FORM_QUORUM.

File: server/src/main/java/org/elasticsearch/cluster/coordination/indicators/CanFormQuorumIndicator.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        return CLUSTER_COORDINATION;
+    }
+
+    @Override
+    public HealthIndicatorResult calculate() {
+        var clusterState = clusterService.state();
+        boolean hasQuorum = clusterState.getLastCommittedConfiguration()
+            .hasQuorum(clusterState.getNodes().getMasterNodes().values().stream().map(DiscoveryNode::getId).collect(Collectors.toSet()));
+        return hasQuorum
+            ? HealthIndicatorResult.of(LACK_OF_QUORUM, CLUSTER_COORDINATION, HealthStatus.GREEN, "Instances can form a quorum.")
```
Comment (tbrooks8) [inline server/src/main/java/org/elasticsearch/cluster/coordination/indicators/CanFormQuorumIndicator.java:47]: I feel like the summary here should be a bit more descriptive. Perhaps "There are enough master-eligible instances to form a quorum."

**Review by Tim-Brooks — COMMENTED** [review]
I added David as a reviewer as I am sure that he has opinions about this type of indicator.

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
My [earlier review](https://github.com/elastic/elasticsearch/pull/82874#discussion_r800664295) still applies here:

We cannot detect whether there are enough nodes to form a quorum by looking only at local state, so a local-only check like this will give false information (both false positives and false negatives). We report local quorum information in the logs today despite this problem, and experience shows that it is indeed something that users find confusing and difficult to interpret.

This check doesn't really make sense at all on non-master nodes because these nodes are not involved in quorum-forming.

Also this is IMO the last thing to check when diagnosing a missing master. It's not a very useful check in practice: if everything else looks fine but there's no master then it's pretty much guaranteed to be a quorum problem. I've seen maybe two exceptions to this since 7.0 was released, both of which were ultimately due to very rare bugs in unrelated components.

I therefore would prefer that we defer work on this indicator until later on and focus on other things for the first release.

**Comment by arteam** [comment]
Let's put this indicator on hold

**Pull request closed without merging**